File: python/design/peeking_iterator.py
class PeekingIterator:
    def __init__(self, iterator):
        def get_elements_in_iterator(iterator):
            ret = []
            while iterator.hasNext():
                ret.append(iterator.next())
            return ret
            
        self.sublist = get_elements_in_iterator(iterator)
        self.pointer = -1
        self.iterator =iterator

    def peek(self):
        return  self.sublist[self.pointer]
        

    def next(self):
        self.pointer+=1
        self.iterator.next()
        return  self.sublist[self.pointer]
        

    def hasNext(self):
        idx = self.pointer+1
        return idx < len(self.sublist)

File: java_leetcode/com/leetcode/PeekingIterator.java
package java_leetcode.com.leetcode;

import java.util.Iterator;
import java.util.LinkedList;

// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html

// 14 / 14 test cases passed.
// Status: Accepted
// Runtime: 8 ms
// Memory Usage: 43.7 MB

public class PeekingIterator implements Iterator<Integer> {
    private final Iterator<Integer> iterator;
    private  LinkedList<Integer> queue;
 	public PeekingIterator(Iterator<Integer> iterator) {
        this.iterator = iterator;
        this.queue =  new LinkedList<>();
        while(iterator.hasNext()){
             queue.add(iterator.next());
        }
    }
	
    // Returns the next element in the iteration without advancing the iterator.
	public Integer peek() {
        return  queue.peekFirst();
	}
	
	// hasNext() and next() should behave the same as in the Iterator interface.
	// Override them if needed.
	@Override
	public Integer next() {
        return queue.pollFirst();
	}
	
	@Override
	public boolean hasNext() {
	    return !queue.isEmpty();
	}
}